fontforge/splinestroke.c

```diff
@@ -32,3689 +32,1805 @@
 #include "baseviews.h"
 #include "cvundoes.h"
 #include "fontforge.h"
+#include "splinefit.h"
 #include "splinefont.h"
 #include "splineorder2.h"
 #include "splineoverlap.h"
 #include "splineutil.h"
 #include "splineutil2.h"
+#include "utanvec.h"
 
+#include <assert.h>
 #include <math.h>
 
-#define PI      3.1415926535897932
-
-typedef struct strokepoint {
-    Spline *sp;
-    bigreal t;
-    BasePoint me;		/* This lies on the original curve */
-    BasePoint slope;		/* Slope at that point */
-    BasePoint left;
-    BasePoint right;
-    /* double radius_of_curvature; */
-    unsigned int butt_bevel: 1;		/* the butt line cap and bevel line join always produce an edge which will be hidden by the pen. We don't want that to happen so must special case them. */
-    unsigned int left_hidden: 1;
-    unsigned int right_hidden: 1;
-    unsigned int hide_left_if_on_edge: 1;
-    unsigned int hide_right_if_on_edge: 1;
-    unsigned int extemum: 1;
-    unsigned int circle: 1;		/* On a cap or a join, point lies on a circular arc centered at me. So slope is the normal to (left-me) */
-    unsigned int line: 1;		/* On a cap, join or place where slope paralel to poly edge, point lies on a line can find slope by vector (pos,pos-1) or (pos+1,pos) */
-    unsigned int needs_point_left: 1;	/* Contour changes direction abruptly on the left side */
-    unsigned int needs_point_right: 1;
-    uint8 lt;
-    uint8 rt;
-} StrokePoint;
-
-/* Square line caps and miter line joins cover more area than the circular pen*/
-/*  So we need a list of the polygons which represent these guys so we can do */
-/*  an additional hit test on them. */
-/* I'm not sure that these will always be presented to us in a clockwise fashion*/
-/*  We can test that by taking a point which is the average of the vertices */
-/*  and seeing if the hit-test says it is inside (It will be inside because */
-/*  these polies are convex). If the hit test fails then the poly must be */
-/*  drawn counter clockwise */
-struct extrapoly {
-    BasePoint poly[4];
-    int ptcnt;		/* 3 (for miters) or 4 (for square caps) */
-};
+#define CIRCOFF 0.551915
 
-enum pentype { pt_circle, pt_square, pt_poly };
+// Rounding makes even slope-continuous splines only approximately so.
+#define INTERSPLINE_MARGIN (1e-1)
```

Comment:
I am guessing these only make sense if we use doubles? Probably more reason to just enforce that doubles are always used and get rid of the configuration option.

Reply:
I pretty much used bigreal everywhere to represent the lack of thought I put into considering what could be lower precision. Getting real and bigreal cleanly separated after years of having both be double would be a mini-project at least, and this code would have to be some part of that project. 

However, on your narrow question, no: INTRASPLINE_MARGIN might well have to change, but INTERSPLINE_MARGIN has to do with int-rounding at smooth points, so it would continue to have that value.